File: core-data/src/main/java/home/oleg/placesnearme/coredata/dao/DetailedVenueHistoryDao.java
package home.oleg.placesnearme.coredata.dao;

import java.util.List;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import home.oleg.placesnearme.coredata.model.DetailedVenueHistory;
import home.oleg.placesnearme.coredata.model.DetailedVenueHistoryDbEntity;
import io.reactivex.Flowable;
import io.reactivex.Maybe;
import io.reactivex.Single;

@Dao
public interface DetailedVenueHistoryDao {
    @Transaction
    @Query("SELECT detailed_venue.*, venue_history.createdAt FROM detailed_venue " +
            "INNER JOIN venue_history ON venue_history.venueId = detailed_venue.id")
    Flowable<List<DetailedVenueHistory>> getAllHistory();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(DetailedVenueHistoryDbEntity detailedVenueHistoryDbEntity);

    @Transaction
    @Query("SELECT detailed_venue.*, venue_history.createdAt FROM detailed_venue " +
            "INNER JOIN venue_history ON venue_history.isLastCheckIn = 1")
    Single<DetailedVenueHistory> getCurrent();

    @Query("SELECT * FROM venue_history WHERE venue_history.venueId = :venueId")
    Flowable<DetailedVenueHistoryDbEntity> observeById(String venueId);

    @Query("SELECT * FROM venue_history WHERE isLastCheckIn = 1")
    Maybe<DetailedVenueHistoryDbEntity> getLastCheckIn();

    @Query("DELETE FROM venue_history WHERE venue_history.venueId = :venueId")
    void remove(String venueId);

    @Query("SELECT * FROM venue_history WHERE venue_history.venueId = :venueId")
    Single<DetailedVenueHistoryDbEntity> getById(String venueId);
}

File: core-data/src/main/java/home/oleg/placesnearme/coredata/dao/DetailedVenueHistoryDao.kt
package home.oleg.placesnearme.coredata.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Transaction
import home.oleg.placesnearme.coredata.database.AppDatabase
import home.oleg.placesnearme.coredata.model.DetailedVenueHistory
import home.oleg.placesnearme.coredata.model.DetailedVenueHistoryDbEntity
import home.oleg.placesnearme.coredata.model.PLACES_HISTORY_TABLE_NAME
import home.oleg.placesnearme.coredata.model.PLACES_TABLE_NAME
import io.reactivex.Flowable
import io.reactivex.Maybe
import io.reactivex.Single

@Dao
abstract class DetailedVenueHistoryDao : CrudDao<DetailedVenueHistoryDbEntity>(PLACES_HISTORY_TABLE_NAME) {
    @Transaction
    @Query("SELECT $PLACES_TABLE_NAME.*, $PLACES_HISTORY_TABLE_NAME.createdAt " +
            "FROM $PLACES_TABLE_NAME " +
            "INNER JOIN $PLACES_HISTORY_TABLE_NAME ON $PLACES_HISTORY_TABLE_NAME.placeId = $PLACES_TABLE_NAME.id")
    abstract fun allHistory(): Flowable<List<DetailedVenueHistory>>

    @Query("SELECT * FROM $PLACES_HISTORY_TABLE_NAME WHERE isLastCheckIn = 1")
    abstract fun lastCheckIn(): Maybe<DetailedVenueHistoryDbEntity>

    @Query("SELECT * FROM $PLACES_HISTORY_TABLE_NAME WHERE $PLACES_HISTORY_TABLE_NAME.placeId = :venueId")
    abstract fun observeById(venueId: String): Flowable<DetailedVenueHistoryDbEntity>

    @Query("DELETE FROM $PLACES_HISTORY_TABLE_NAME WHERE $PLACES_HISTORY_TABLE_NAME.placeId = :venueId")
    abstract fun remove(venueId: String)

    @Query("SELECT * FROM $PLACES_HISTORY_TABLE_NAME WHERE $PLACES_HISTORY_TABLE_NAME.placeId = :venueId")
    abstract fun getById(venueId: String): Single<DetailedVenueHistoryDbEntity>

    @Transaction
    @Query("SELECT $PLACES_TABLE_NAME.*, $PLACES_HISTORY_TABLE_NAME.createdAt FROM $PLACES_TABLE_NAME" +
            " INNER JOIN $PLACES_HISTORY_TABLE_NAME ON $PLACES_HISTORY_TABLE_NAME.isLastCheckIn = 1")
    abstract fun current(): Single<DetailedVenueHistory>

    @Transaction
    open fun checkIn(placeId: String) {
        val historyDbEntity = DetailedVenueHistoryDbEntity(
                createdAt = System.currentTimeMillis(),
                isLastCheckIn = true,
                placeId = placeId
        )
        dropCheckIns()
        insertOrReplace(historyDbEntity)
    }

    @Query("UPDATE $PLACES_HISTORY_TABLE_NAME SET isLastCheckIn = 0 WHERE isLastCheckIn = 1")
    abstract fun dropCheckIns() : Long

}
